presentation/src/main/java/org/sagebionetworks/research/presentation/speech/TextToSpeechService.kt

```diff
@@ -0,0 +1,420 @@
+                                LOGGER.info("There is more text to speak, updating state to QUEUED")
+                                // More speech is queued
+                                _state.postValue(TextToSpeechState(QUEUED, null))
+                            }
+                        }
+
+                        override fun onError(text: String?) {
+                            _state.postValue(TextToSpeechState(ERROR, null))
+                        }
+
+                        override fun onStart(text: String?) {
+                            _state.postValue(TextToSpeechState(SPEAKING, text))
+                        }
+                    })
+                }
+
+                textToSpeakOnInit?.let { speakText(it) }
+            } else {
+                LOGGER.warn("Language not available for TTS")
+                textToSpeech?.shutdown()
+                textToSpeech = null
+            }
+        } else {
+            LOGGER.warn("Failed to initialize TTS with error code $status")
+            textToSpeech = null
+        }
+    }
+
+    /**
+     * Register the given speech map to be spoken at the times determined by the given countdown and duration.
+     * @param duration the duration of the current step
+     * @param countdown the countdown LiveData for the current step
+     * @param speechMap the map where keys are offsets into the countdown and values are instructions that should be
+     * spoken to the user.
+     */
+    fun registerSpeechesOnCountdown(duration: Duration, countdown: LiveData<Long>, speechMap: Map<String, String>) {
+        clear()
+        LOGGER.debug("register speeches on countdown called,\nduration: " +
+                "$duration\ncountdown: $countdown\nspeechMap: $speechMap")
+        textToSpeech = TextToSpeech(this, this)
+        val canonicalSpeechMap = getCanonicalSpeechMap(speechMap, duration).toMutableMap()
+        val currentCount = countdown.value
+        if (currentCount != null) {
+            // filter out all the speeches that shouldn't be spoken right now, and sort the ones that should in
+            // increasing order of offset.
+            val skippedSpeeches = canonicalSpeechMap.filter { entry ->
+                val currentOffset = duration.seconds - currentCount
+                entry.key < currentOffset && currentOffset - entry.key < speechGracePeriod
+            }.toSortedMap()
+
+            for (speech in skippedSpeeches) {
+                if (!speech.value.second) {
+                    LOGGER.debug(
+                            "Spoken Instruction: {}, should have been spoken before the service was initialized," +
+                                    " and was still within the grace period.")
+                    speakText(speech.value.first)
+                    canonicalSpeechMap[speech.key] = speech.value.copy(second = true)
+                }
+            }
+        }
+        val countdownObserver = Observer<Long> {
```

Comment:
@Robert-Kolmos @liujoshua this is important to know...

This was bug was causing the most issues. The observer wasn't ever being removed.  
There was this line of code that was supposed to remove it, but it wasn't actually working...
`futureSpeechData?.countdown?.removeObserver(:: onCountdownChanged)`

Why wasn't this removing the observer since we called observeForever in the exact same way? From tracking it down within the LiveData class itself, it's because lambda expressions that convert object function (this::onCountdownChanged) to Observables automatically, don't have reproducible hashCodes.  Java generates a new Observable every time the (this::onCountdownChanged) is referenced.  

So, when I stored the observer in FutureSpeechData, and didn't try and reference it again with (this::onCountdownChanged), the LiveData was able to find it in it's observable map and remove it correctly.